### PR TITLE
Gobblin Dashboard for jobs

### PR DESCRIPTION
**Pull request by eogren**
- Add a UI that queries the JobExecutionInfoServer for various job metrics and displays it in an easy to read format. If `admin.server.enabled` is set to true, this will be served on port 8000
 - A Cli tool is also provided, which can easily be run with the bin/gobblin-admin.sh script
 - Few changes to the JobExecutionInfoServer to retrieve the data we want

Thanks to @ahollenbach for doing most of the work!

### CODE REVIEW

**Comment by ahollenbach** [comment]
Good changes 😊👍

**Comment by chavdar** [comment]
@abti and @ibuenros Can you please review?

**Comment by ibuenros** [comment]
@eogren The PR is quite large. Any suggestion on the order in which we should read the new classes? A start point? Thanks!

**Comment by eogren** [comment]
Hey @ibuenros - sure, and sorry for the gigantic commit. It was tough to squash it into a bunch of smaller ones.

I would recommend starting with the gobblin-rest-service/ and gobblin-metastore/ changes which are the API changes necessary to power the dashboard.

Next, optionally, is probably gobblin-admin/.../Cli.java to get a sense of some of the queries we may be exposing.

The web files in gobblin-admin/ next - although the web UI is really just making calls to the Gobblin REST server and turning them into tables and fancy graphs.

Last is probably the changes to SchedulerDaemon and the AdminWebServer -- these just start an embedded Jetty server that serves all the web stuff to make it easier for people to get started with the dashboard.

Hope that helps

File: gobblin-admin/src/main/java/gobblin/cli/Cli.java [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+
+/**
+ * A command line interface for interacting with Gobblin.
+ * From this tool, you should be able to:
+ *  * Check the status of Gobblin jobs
+ *  * View ...
+ *
+ * @author ahollenbach@nerdwallet.com
+ */
+public class Cli {
```
Comment (ibuenros) [inline gobblin-admin/src/main/java/gobblin/cli/Cli.java]: This class is too long atm. Can you split it? Suggestion: Have classes `JobCommand`, and a class `TaskCommand`.

File: gobblin-admin/src/main/java/gobblin/cli/Cli.java [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+  private static enum Command {
```
Comment (ibuenros) [inline gobblin-admin/src/main/java/gobblin/cli/Cli.java]: Suggestion for alternative way of doing this:

```java
  private static enum Command {
    JOBS(JobCommand.getClass()), 
    TASKS(TaskCommand.getClass());

    private Class<? extends Command> commandClass;

    Command(Class<? extends Command> commandClass) {
      this.commandClass = commandClass;
    }

    public Command getCommandObject(CommandLine parsedOpts) {
      return ConstructorUtils.invokeConstructor(this.commandClass, parsedOpts);
    }
  }
```

then in `execute()` you can do:
```
Command.valueOf(this.command.toUpperCase()).getCommandObject(parsedOpts).execute();
```

IMO it makes it easier to add more commands in the future.

I would also move `Command` enum to a separate top level class.

File: gobblin-admin/src/main/java/gobblin/cli/Cli.java [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+
+  private void executeTaskCommand() {
+    throw new NotImplementedException("Tasks lookup not yet implemented!");
+  }
+
+  private boolean shouldExecuteCommand(Command cmd) {
+    return this.command.equals(cmd.name().toLowerCase());
+  }
+
+  private JobExecutionInfo queryByJobId(String id, boolean includeProperties) {
```
Comment (ibuenros) [inline gobblin-admin/src/main/java/gobblin/cli/Cli.java]: Is it intentional that you are not using `includeJobProperties`?

File: gobblin-admin/src/main/java/gobblin/cli/Cli.java [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+    new CliTablePrinter.Builder()
+        .labels(labels)
+        .data(data)
+        .flags(flags)
+        .delimiterWidth(2)
+        .build()
+        .printTable();
+  }
+
+  private JobExecutionInfo queryJobProperties(String id) {
```
Comment (ibuenros) [inline gobblin-admin/src/main/java/gobblin/cli/Cli.java]: This method is never used, but may be related to the above comment.

File: gobblin-admin/src/main/java/gobblin/cli/CliTablePrinter.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+public class CliTablePrinter {
+  /**
+   * Labels for each columns
+   */
+  private List<String> labels;
+
+  /**
+   * A list of flag strings (corresponding to each column)
+   */
+  private List<String> flags;
```
Comment (ibuenros) [inline gobblin-admin/src/main/java/gobblin/cli/CliTablePrinter.java]: Can you explain a bit more what flags mean? I see now they are the flags used for String formatting, but it wasn't clear at the beginning.

**Comment by eogren** [comment]
Sorry for the delay - I split the Cli stuff into a bunch of smaller classes using a similar (although not exactly the same) method as proposed. Unfortunately it meant basically rewriting the Cli class so a lot of the original context may be lost.

**Comment by ibuenros** [comment]
LGTM

**Comment by ibuenros** [comment]
Can you make a wiki entry on the admin server?

**Comment by ibuenros** [comment]
@abti can you take a final look at the PR before merging? Thanks.

**Comment by abti** [comment]
@ibuenros sure I will go through it later today

**Pull request merged**